Test failures on Python 3.6.0b2

They seem to result from nesting `warnings.catch_warnings` and `suppress_warnings`. The problem is new in 3.6 and may be either be a Python bug, or something we are doing in `suppress_warnings` that we shouldn't.
```
FAIL: test_deprecations.TestComparisonDeprecations.test_array_richcompare_legacy_weirdness
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_deprecations.py", line 371, in test_array_richcompare_legacy_weirdness
    assert_warns(FutureWarning, lambda: np.arange(2) == "a")
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1723, in assert_warns
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1680, in _assert_warns_context
    raise AssertionError("No warning raised" + name_str)
AssertionError: No warning raised when calling <lambda>

======================================================================
FAIL: test_deprecations.TestComparisonDeprecations.test_none_comparison
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_deprecations.py", line 315, in test_none_comparison
    assert_warns(FutureWarning, operator.eq, np.arange(3), None)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1723, in assert_warns
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1680, in _assert_warns_context
    raise AssertionError("No warning raised" + name_str)
AssertionError: No warning raised when calling eq

======================================================================
FAIL: test_deprecations.TestComparisonDeprecations.test_void_dtype_equality_failures
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_deprecations.py", line 359, in test_void_dtype_equality_failures
    assert_warns(FutureWarning, lambda: struct1 == 1)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1723, in assert_warns
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1680, in _assert_warns_context
    raise AssertionError("No warning raised" + name_str)
AssertionError: No warning raised when calling <lambda>

======================================================================
FAIL: test_deprecations.TestIdentityComparisonDeprecations.test_identity_equality_mismatch
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_deprecations.py", line 415, in test_identity_equality_mismatch
    assert_warns(FutureWarning, np.equal, a, a)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1723, in assert_warns
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1680, in _assert_warns_context
    raise AssertionError("No warning raised" + name_str)
AssertionError: No warning raised when calling equal

======================================================================
FAIL: test_deprecations.TestRankDeprecation.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/core/tests/test_deprecations.py", line 272, in test
    assert_warns(np.VisibleDeprecationWarning, np.rank, a)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1723, in assert_warns
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 1680, in _assert_warns_context
    raise AssertionError("No warning raised" + name_str)
AssertionError: No warning raised when calling rank

======================================================================
FAIL: test_utils.test_suppress_warnings_decorate_no_record
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/charris/.local/lib/python3.6/site-packages/nose-1.3.6-py3.6.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/tests/test_utils.py", line 945, in test_suppress_warnings_decorate_no_record
    assert_(len(w) == 1)
  File "/home/charris/Workspace/numpy.git/build/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 90, in assert_
    raise AssertionError(smsg)
AssertionError

----------------------------------------------------------------------
```

--- a/numpy/testing/utils.py
+++ b/numpy/testing/utils.py
@@ -2027,6 +2027,11 @@ class suppress_warnings(object):
         self._forwarding_rule = forwarding_rule
 
     def _clear_registries(self):
+        if hasattr(warnings, "_filters_mutated"):
+            # clearing the registry should not be necessary on new pythons,
+            # instead the filters should be mutated.
+            warnings._filters_mutated()
+            return
         # Simply clear the registry, this should normally be harmless,
         # note that on new pythons it would be invalidated anyway.
         for module in self._tmp_modules:
@@ -2116,6 +2121,8 @@ class suppress_warnings(object):
             raise RuntimeError("cannot enter suppress_warnings twice.")
 
         self._orig_show = warnings.showwarning
+        if hasattr(warnings, "_showwarnmsg"):
+            self._orig_showmsg = warnings._showwarnmsg
         self._filters = warnings.filters
         warnings.filters = self._filters[:]
 
@@ -2139,20 +2146,29 @@ class suppress_warnings(object):
                     module=module_regex)
                 self._tmp_modules.add(mod)
         warnings.showwarning = self._showwarning
+        if hasattr(warnings, "_showwarnmsg"):
+            warnings._showwarnmsg = self._showwarnmsg
         self._clear_registries()
 
         return self
 
     def __exit__(self, *exc_info):
         warnings.showwarning = self._orig_show
+        if hasattr(warnings, "_showwarnmsg"):
+            warnings._showwarnmsg = self._orig_showmsg
         warnings.filters = self._filters
         self._clear_registries()
         self._entered = False
         del self._orig_show
         del self._filters
 
+    def _showwarnmsg(self, msg):
+        self._showwarning(msg.message, msg.category, msg.filename, msg.lineno,
+                          msg.file, msg.line, use_warnmsg=msg)
+
     def _showwarning(self, message, category, filename, lineno,
                      *args, **kwargs):
+        use_warnmsg = kwargs.pop("use_warnmsg", None)
         for cat, _, pattern, mod, rec in (
                 self._suppressions + self._tmp_suppressions)[::-1]:
             if (issubclass(category, cat) and
@@ -2179,8 +2195,11 @@ class suppress_warnings(object):
         # There is no filter in place, so pass to the outside handler
         # unless we should only pass it once
         if self._forwarding_rule == "always":
-            self._orig_show(message, category, filename, lineno,
-                            *args, **kwargs)
+            if use_warnmsg is None:
+                self._orig_show(message, category, filename, lineno,
+                                *args, **kwargs)
+            else:
+                self._orig_showmsg(use_warnmsg)
             return
 
         if self._forwarding_rule == "once":
@@ -2193,7 +2212,11 @@ class suppress_warnings(object):
         if signature in self._forwarded:
             return
         self._forwarded.add(signature)
-        self._orig_show(message, category, filename, lineno, *args, **kwargs)
+        if use_warnmsg is None:
+            self._orig_show(message, category, filename, lineno, *args,
+                            **kwargs)
+        else:
+            self._orig_showmsg(use_warnmsg)
 
     def __call__(self, func):
         """

